IoUring: use typed Flags and start syncing with liburing 2.12

bernardassan opened the pull request:
# IoUring Overhaul

- [x] Use Named Constants for `IoUring`
- [x] Move `io_uring_sqe.zig` into `IoUring`
- [x] Add newly exposed `IoUring` Types, Constants, and Operations
- [x] Improve api of `init` fn with `Flags.Setup`
__Incompatible Flags References__
https://github.com/axboe/liburing/issues/1075#issuecomment-1970744640
https://github.com/axboe/liburing/issues/811#issuecomment-1454967969
https://nick-black.com/dankwiki/index.php/Io_uring
- [x] Catch simple incompatible flag configurations during `init_params` fn
- [x] Improve naming and namespacing
- [x] Update IoUring to match [liburing 2.12](https://github.com/axboe/liburing/blob/liburing-2.12/src/include/liburing/io_uring.h)
- [x] follow Zig [Style Guide](https://ziglang.org/documentation/master/#Style-Guide) where it makes sense
- [x] Ensure there are no regressions (It would be great if community members already using it can test it out and give feedback)
- [x] Remove IoUring bits and pieces from `linux.zig`
- [x] Update IoUring to use the new typed flags
- [x] Add `int_flags` to fix a real-world issue in ghostty https://github.com/ziglang/zig/issues/25566
- [x] Cleanup `linux/test.zig` and `IoUring` tests
- [x] Update the added flags to match options available in kernel 6.16 

## Future Enhancement
- Indicate what versions of the kernel support the new flags and operations
- Port all tests from liburing 2.12 to IoUring
- Implement functionalities that require `int_flags` and `enter_ring_fd`

bernardassan committed:
IoUring: use typed Flags and Features for IoUring

IoUring Overhaul

Named Constants for IoUring

Move io_uring_sqe.zig into IoUring

Add new exposed IoUring data types

Improve naming and namespacing a bit

Improve api of init with Setup.Flags

 // https://github.com/axboe/liburing/issues/1075#issuecomment-1970744640
 // https://github.com/axboe/liburing/issues/811#issuecomment-1454967969
 // https://nick-black.com/dankwiki/index.php/Io_uring
Catch simple incompatible flag configurations during init_params

Start updating IoUring gradually to use the new typed flags

Slowly remove IoUring bits and pieces from linux.zig

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
replace some more fn flags with Typed Flags

and use linux Flags where appropriate

constants is a namespace

make AsyncCancel flags more descriptive

move cqe.flags into the Cqe type

make io_uring_sqe ioprio a packed union of SendRecv & Accept

TODO: move some of the used linux Flags to be packed structs

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
move Flags that were mistakenly tagged as constants

Improve naming and namespacing of IoUring flags and types to
match Zig Style

move some flags from uflags to its type

Add matches doc comments to make it easy to match Zig IoUring types
to liburing

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Replace STATX_* with StatxMask & StatxAttr

Add RenameFlags for IoUring renameat

Start update io_uring Types in linux.zig to use IoUring

Continue change options and flags to Typed types

Fix test cases to use updated IoUring flags and options

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Improve organization of fn and structs in IoUring

Update std to use Statx.Mask
Update statx syscall to use Statx.Mask

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Replace AT,W,SHUT,SOCK with a packed struct Flag type

Use the new Flag types in IoUring

fix some test

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
update some syscall APIs to use the new flags

Add Shut, PF, AF, IPPROTO types

Update IoUring and test to use some of this flags

TODO: avoid breaking API and make transitioning smooth
Instead mark the previous API as deprecated for at least
One zig release while using the new in the Zig codebase

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Restore deprecated contants using new Flag types

This allows for a gradual transitioning to the new typed flags

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Replace MSG with Packed Struct Flags

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Get test passing for all the newly introduced flags

4❯ zig test lib/std/std.zig --zig-lib-dir lib --test-filter "os.linux"
68/116 os.linux.test.test.futex2_wait...SKIP
69/116 os.linux.test.test.futex2_wake...SKIP
70/116 os.linux.test.test.futex2_requeue...SKIP
111/116 os.linux.IoUring.test.waitid...SKIP
116/116 os.linux.IoUring.test.bind/listen/connect...SKIP
111 passed; 5 skipped; 0 failed.

fix skipKernelLessThan fn to work on wsl linux

fix logical error in flags compatibility check

Update all IoUring to use the implemented flags

bandage types that affect posix for now so that the scope
doesn't become unmanageable

TODO: impl linux.SOL and linux.SO types

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Add So and Sol typed flags

Replace SO with So type and Add SO named constants for compatability

Replace SOL with Sol type and Add SOL named constants for compatability

Replace some more typed flags with their typed equivalent

Remove any unnecessary IoUring dependency on linux.*

Add TODOs for all untyped flags

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Replace EPOLL struct with an EpollOp enum and Epoll packed struct type

Prefer Using aliases over packed union field

Fix bug in HANDLE_FID declaration

Use EpollOp op type and Epoll packed struct flag type in IoUring

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
fix error of setting some fields in Epoll to true by default

Add Deprecated Op constants for compatibility

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Remove io_uring_sqe.zig from CMakeLists

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Use lower case identifiers for IoUring flags and enums

update io_uring_enter and io_uring_register syscalls to
use IoUring flags and types

Cleanup comments

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
replace direct set of some flags with link_next and set_flags calls

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Remove io_uring bit and pieces from linux.zig

now that IoUring.zig is self contained
Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Fix posix.W in process/Child.zig

add Deprecated linux.W fns
Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Add mips defination for Epoll

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Add and improve comments

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Remove unnecessary use of @as coercion

update `cq_ring_needs_flush()` to match liburing

replace FilesUpdate with RsrcUpdate struct
Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Implement more IoUring register functions

Future enhancement:
enter_ring_fd: u24,
init_flags: u8,

register functions that require the extra fields above like
io_uring_register_ring_fd, io_uring_close_ring_fd,
io_uring_clone_buffers_offset, io_uring_set_iowait not implemented

register_*_tag functions not implemented

remove unnecessary use of @ptrCast in io_uring_register

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Remove unnecessary null to optional anyopaque coercion

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Move buf_ring_* functions into BufferRing type as methods

add IoUring tee syscall functionality

adapt ReadBuffer union idea for read_fixed to support fixed read and readv

Use same idea and create a WriteBuffer type for write() and write_fixed()

ignored flags for splice and tee lets see if they become important in
the future

add a get_data helper to Cqe

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Implement some more IoUring operations

Add timeout update, cancel_fd, epoll_wait, files_update, open, open_direct, f/madvice

don't implement f/madvice64 for now I doubt it is used by a lot of people in practice

Implement BufferGroup read_multishot

Try to make definition of function align as logically as posible with
liburing to make keeping our implementation in sync easy

TODO: add M/Fadvice enum flags
TODO: understand buffer_selection and how its different from BufferGroup

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
add IoUring send_bundle, send_to, recv_multishot, sync_file_range

ignore prep_openat2* for now

add non-at variant of some unlink, rename, mkdir, symlink, link

TODO: add Sync File Flags

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
add msg_ring_*, setxattr and getxattr IoUring operations

TODO: add SetXattr flags

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

RaidoAun commented:
There currently does not seem to be a replacement for int_flags on the ring. liburing uses that to set NO_IOWAIT flag to then on enter instead of initializing flags as 0 it is instead set to a mask of int_flags which includes NO_IOWAIT. 
see:
https://github.com/ziglang/zig/pull/25604

and:
https://github.com/axboe/liburing/commit/8f0c124785f167a6fbe034e59c32b63a4caea702#diff-ddb623c7d8d649eff38f34b19b7a2f8b55e1d448ee94b813a9541d5304467629

otherwise I like these changes since they add type safety around the flags. While playing with this I found the bug in my own PR.

bernardassan commented:
This is on my to-do list, but I wanted to first get `IoUring` up-to-date with `liburing 2.12` before working on these changes. Because I realized some liburing operations (Check commit message https://github.com/ziglang/zig/pull/25394/commits/6bb7f304bf719dc3e748f4eca1e68e54802f664e) depend on `int_flags` and `enter_ring_fd`. 

But because this is affecting a real-world projects [ghostty](https://github.com/ziglang/zig/issues/25566) which I use :smile: , let me add that functionality here quickly so you can test it out for me.

bernardassan committed:
IoUring: Implement set_iowait functionality

closes https://github.com/ziglang/zig/issues/25566
closes https://github.com/ziglang/zig/pull/25604

and IoUring enter_flags() helper function
add typed Init flags which match int_flags in liburing

Co-authored-by: RaidoAun <45874591+RaidoAun@users.noreply.github.com>
Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan commented:
@RaidoAun a port of your solution in https://github.com/ziglang/zig/pull/25604 is ready in this branch

And can you share your GitHub email so I add you as a Co-Author of that commit

RaidoAun commented:
Thanks, I will take a look.

Im fine without being co authored, but if you insist then you can just use my public github email (45874591+RaidoAun@users.noreply.github.com). It should also be visible on the commit on my PR.

RaidoAun commented on lib/std/os/linux/IoUring.zig:
```diff
@@ -298,89 +342,122 @@ fn copy_cqes_ready(self: *IoUring, cqes: []linux.io_uring_cqe) u32 {
+        self.init_flags.toggle_iowait();
+    } else {
+        self.init_flags.no_iowait = true;
+    }
```
pretty sure this whole if can be removed and use this instead:
self.init_flags.no_iowait = !enable_iowait

then the toggle_iowait function could also be removed.

RaidoAun commented on lib/std/os/linux/IoUring.zig:
```diff
@@ -1670,217 +3426,1099 @@ pub const BufferGroup = struct {
+        _9: u24 = 0,
+
+        /// check that flags that match flags in `Init` aren't set
+        pub fn no_init_flags(flags: Enter) bool {
```
if this function is called with empty flags, then it passes. If this is called with some invalid flag being set then this will still pass (returns true). this should instead check if any other flags not including the no_iowait or registered flags are set. In other words this should ignore the two flags and if any others are set then it should indicate a problem.

from my PR:
`assert(flags.* & (~@as(u8, linux.IORING_INT_FLAGS_MASK)) == 0);`
this should be equivalent with that.

regarding the naming I would not use a negative for the function name, I would propose renaming this to "valid_init_flags" or "only_allowed_init_flags". then assertions would be: assert(!only_allowed_init_flags)
Or make this function itself include an assert and have this just be: "assert_valid_init_flags"

already the flag in the kernel (no_iowait) is causing me a headache with the naming having a negative in it.

bernardassan replied:
I like the `valid_init_flags` better, so I'm going with that.

bernardassan committed:
Add XattrSource to decide how to prepare set/getxattr operations

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

RaidoAun reviewed (approved):
The code regarding NO_IOWAIT looks fine now. If I find time to experiment with io_uring then I will try to test more of this branch. This branch has gotten quite big and I think you should try to get a core team members opinions on this. I would really like to see these changes get merged. 
If you need help with this then I am willing to also contribute.

bernardassan commented:
That is true, I didn't expect it to get this big, but I guess because we are way behind `liburing 2.12` that should have been expected. @RaidoAun, a helping hand is always welcome. It would be nice to have an extra hand on deck when working on the future enhancements stated in the PR description.

I agree it's a good time to start getting reviews from core team members and previous contributors, as there are only a few more operations I want to implement and leave the rest as future enhancements, which I would make another PR for.

bernardassan commented:
cc @mlugg @ianic @andrewrk @nikneym @topecongiro @ozgrakkurt @psnszsn @joadnacer @Vexu @jacobly0 @mllken @vrischmann @hnakamur @lithdew @ayende @jorangreef @ifreund @daurnimator

bernardassan commented:
**On my Arch Linux machine**

❯ zig test lib/std/std.zig --zig-lib-dir lib --test-filter "os.linux"
All 117 tests passed.

bernardassan committed:
IoUring: futex operations

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

Add futex_* operations to IoUring

Refactor Futex 2 flags into Futex2 struct

add mpol to Wait flags and fix private field as its 128 not 32

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

Update futex2_* functions and constants to use the new Futex2 type

Improve the Api of futex2_* functions to be more idiomatic Zig

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Add a Futex2 `Bitset` type for futex2 wake and wait syscalls

Cleanup linux/test.zig

type futex/2_* uaddr as *const u32
consider changing to *const atomic.Value(u32)

Use At Flags in fstatat

Use EpollOp in epoll_ctl syscall

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
IoUring: Working on Pipe2 flags

Build on the extensive work already done

IoUring: add fixed_fd_install, ftruncate, cmd_discard

Working on IoUring pipe flags

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
IoUring: implement pipe and pipe_direct operations

Add dedicated Pipe2 flags for pipe2 as O contains some fields which are
invalid in that context

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Merge archs that have the same Pipe2 flags

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Use linux types directly since IoUring is only supported on linux

Use explicit array types were possible

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
IoUring: implement outstanding flags and enumerations

Add SYNC_FILE_RANGE_* flags
add XATTR_* flags
Add Madvise for MADV_* constants
Add Fadvise enum type for POSIX_FADV_* constants

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
IoUring: use the splice flags type for splice and tee

Use appropriately sized integers where applicable

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
IoUring: fix and remove TODOs

backout from updating `pipe2` syscall to use Pipe2 flags

Review if useful to implement
- openat2*, f/madvice64

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
IoUring: Fix merge conflicts

use flags in Threaded

change futex signature to expect *const atomic.Value(u32)

fix IoUring timeout test

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Fix `At` flags for real

The fix got reverted in the merge conflict but now properly fix `At`

Update posix `At` to use packed struct (I was holding back from touching
`posix` and `c` but I might as well just get this done instead of using
the deprecated constants since the diff is already big and At is worth
while)

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan committed:
Implement W as packed struct in `c.zig`

fixes the below on CI

/lib/std/process/Child.zig:526:24: error: cannot @bitCast to 'c.W__struct_316926'; struct does not have a guaranteed in-memory layout
    const w: posix.W = @bitCast(status);
                       ^~~~~~~~~~~~~~~~
/lib/std/c.zig:3682:75: note: struct declared here
    .driverkit, .ios, .maccatalyst, .macos, .tvos, .visionos, .watchos => struct {
                                                                          ^~~~~~

Signed-off-by: Bernard Assan <mega.alpha100@gmail.com>

bernardassan commented:
Rebased onto `master`

alexrp commented:
https://codeberg.org/ziglang/zig/pulls/30067
